[lldb][swift][tsan] Strawman fix for TSAN  tests failing to  compile

wrotki opened the pull request:
The tests only refer to swift dylib directory in build, and it doesn't contain the TSAN dylib.

Added the directory containing TSAN dylib, from the LLVM build - however could only locate this LLVM directory by walking backwards from CLANG_MODULE_CACHE_DIR , which I'm told is not reliable enough.

Putting out this PR to discuss potential better solutions.

rdar://103001321

wrotki commented:
@swift-ci please test

adrian-prantl commented on lldb/test/API/CMakeLists.txt:
```diff
@@ -41,7 +41,7 @@ if(LLDB_TEST_SWIFT)
   set(SWIFT_TEST_ARGS
-    --inferior-env "DYLD_LIBRARY_PATH=\\\"${LLDB_SWIFT_LIBS}/macosx\\\""
+    --inferior-env "DYLD_LIBRARY_PATH=\\\"${LLDB_SWIFT_LIBS}/macosx:${LLVM_LIBRARY_OUTPUT_INTDIR}/lldb/clang/lib/darwin\\\""
     --inferior-env "LD_LIBRARY_PATH=\\\"${LLDB_SWIFT_LIBS}/${CMAKE_SYSTEM_PROCESSOR}\\\""
```
@JDevlieghere @bulbazord 

This really surprises me. Why does our CMake file hardcode macOS-specific test args? I guess the answer is because the DYLD_* options don't have any effect on Linux? Still this seems wrong...

bulbazord replied:
> because the DYLD_* options don't have any effect on Linux?

That would be my guess as well. Notice that on the line below we're also passing `LD_LIBRARY_PATH` (even on macOS). We could probably divide these options up based on the host platform if we want to.

adrian-prantl commented on lldb/test/API/functionalities/tsan/swift/Makefile:
```diff
@@ -1,6 +1,9 @@
 SWIFT_SOURCES := main.swift
 
+TSAN_DYLIB_DIR := $(CLANG_MODULE_CACHE_DIR)/../../../lib/lldb/swift/clang/lib/darwin/
+LD_EXTRAS := -lclang_rt.tsan_osx_dynamic -L$(TSAN_DYLIB_DIR)
```
Should we modify `Makefile.rules` such that all tests use the just-built `compiler-rt`?

bulbazord replied:
I think this is a good idea, but we should probably make sure we're consistent and confident about finding the right directory.

adrian-prantl commented on lldb/test/API/functionalities/tsan/swift/Makefile:
```diff
@@ -1,6 +1,9 @@
 
 SWIFT_SOURCES := main.swift
 
+TSAN_DYLIB_DIR := $(CLANG_MODULE_CACHE_DIR)/../../../lib/lldb/swift/clang/lib/darwin/
```
We need to find a way to not derive this from CLANG_MODULE_CACHE_DIR.

wrotki replied:
Right - I dumped all variables available in this makefile, and LLVM_LIBS_DIR one seems better for this purpose. I'll update it, but want to see first what the discussion here is going to be like, perhaps we'll go some other way.

bulbazord replied:
The issue with using `LLVM_LIBS_DIR` is that it points to `$LLVM_BUILD_DIR/lib` instead of `$LLDB_BUILD_DIR/lib`. That means that `TSAN_DYLIB_DIR` would be `$(LLVM_LIBS_DIR)/clang/$clang_version/lib/darwin`, which would mean you would have to find and manage `$clang_version` yourself.

It might be worth modifying the test suite to make `LLDB_LIBS_DIR` available to the make rules so you can take advantage of that. That way we can use something like `$(LLDB_LIBS_DIR)/lldb/clang/lib/darwin` or something to that effect. What do you think?

bulbazord replied:
Another idea: If you know the compiler you are using, you could ask it for its resource directory with something like `$CXX -print-resource-dir` and then append `/lib/darwin` yourself. We do that elsewhere in the test suite. See `find_sanitizer_runtime` in `lldb/test/API/lit.cfg.py`.

wrotki replied:
I noticed that LLDB_LIBS_DIR is actually set in lit.site.cfg.py.in, but not set into environment by lit.cfg.py - so I added it.

wrotki replied:
> This idea seems good to me. A few questions:
> 1.) Will this need to be supported moving forward or just for this branch? If moving forward, you'll want to cherry-pick this to `next`.
> 2.) Is this specific to swift or does this need to happen for all TSAN/ASAN tests?

1.) I think it should be supported. We had previous issues with these tests failing to build on rebranch which this fix should address, to eliminate similar problems in the future. Yes, I will cherry-pick to next.
2.) These two tests (running ASANified/TSANinfied Swift code with LLDB API) are very specific, I don't currently know about any similar cases elsewhere.

adrian-prantl commented:
Thanks @wrotki! I added a couple more people to the review to discuss the best path forward for this.

wrotki committed:
The tests only refer to swift dylib directory in build, and it doesn't contain the TSAN dylib.

Added the directory containing CLANG runtime dylibs, from the LLVM build

Putting out this PR to discuss potential better solutions.

rdar://103001321

wrotki commented:
Pushed another iteration of this, defining the clang-rt dylibs location in Makefile.rules . Also added correction for also failing ASAN/Swift/lldb test case.

bulbazord commented:
This idea seems good to me. A few questions:
1.) Will this need to be supported moving forward or just for this branch? If moving forward, you'll want to cherry-pick this to `next`.
2.) Is this specific to swift or does this need to happen for all TSAN/ASAN tests?

bulbazord reviewed (approved):
I'm happy with this. :)

adrian-prantl commented:
@swift-ci test

wrotki commented:
Not sure why GitHub attached my reply to bulbazord quests to Adrian's conversation. Pasting it here again so it's not missed:

| This idea seems good to me. A few questions:
| 1.) Will this need to be supported moving forward or just for this branch? If moving forward, you'll want to cherry-pick this to next.
| 2.) Is this specific to swift or does this need to happen for all TSAN/ASAN tests?
1.) I think it should be supported. We had previous issues with these tests failing to build on rebranch which this fix should address, to eliminate similar problems in the future. Yes, I will cherry-pick to next.
2.) These two tests (running ASANified/TSANinfied Swift code with LLDB API) are very specific, I don't currently know about any similar cases elsewhere.

wrotki commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

bulbazord commented:
@swift-ci test

wrotki commented:
@swift-ci test

wrotki commented:
The compiler problem which was blocking me from validating this PR is gone now, and the tests failing now are also failing in same way in unmodified stable/20221013 branch, so I think this change should be good to go. Going to merge this.
